app/src/main/java/com/example/campusguide/directions/ChooseOriginOptions.kt

```diff
@@ -0,0 +1,92 @@
+package com.example.campusguide.directions
+
+import android.Manifest
+import android.app.Activity
+import android.graphics.Color
+import android.graphics.drawable.ColorDrawable
+import android.location.Location
+import android.os.Bundle
+import android.view.LayoutInflater
+import android.view.View
+import android.view.ViewGroup
+import android.widget.Button
+import androidx.fragment.app.DialogFragment
+import com.example.campusguide.Constants
+import com.example.campusguide.MapsActivity
+import com.example.campusguide.R
+import com.example.campusguide.location.FusedLocationProvider
+import com.example.campusguide.search.CustomSearch
+import com.example.campusguide.search.SearchLocation
+import com.example.campusguide.search.SearchLocationListener
+import com.example.campusguide.search.indoor.BuildingIndexSingleton
+import com.example.campusguide.search.indoor.IndoorLocationProvider
+import com.example.campusguide.search.indoor.IndoorSearchResultProvider
+import com.example.campusguide.search.outdoor.PlacesApiSearchLocationProvider
+import com.example.campusguide.utils.permissions.PermissionsSubject
+import java.lang.NullPointerException
+
+class ChooseOriginOptions(
+    private val permission: PermissionsSubject,
+    private val locationProvider: FusedLocationProvider,
+    private val locationSelectedListener: (location: Location) -> Unit
+) : DialogFragment() {
+
+    private val locationPermission = Manifest.permission.ACCESS_FINE_LOCATION
+
+    override fun onCreateView(
+        inflater: LayoutInflater,
+        container: ViewGroup?,
+        savedInstanceState: Bundle?
+    ): View? {
+        dialog?.window?.setBackgroundDrawable(ColorDrawable(Color.TRANSPARENT))
+        val view = inflater.inflate(R.layout.choose_origin_options, container, false)
+
+        view.findViewById<Button>(R.id.currentLocation)?.setOnClickListener {
+            useCurrentLocation()
+        }
+
+        view.findViewById<Button>(R.id.search)?.setOnClickListener {
+            searchForLocation()
+        }
+
+        return view
+    }
+
+    /**
+     * Gets the user's last known most recent location
+     */
+    private fun useCurrentLocation() {
+        if (permission.havePermission(locationPermission)) {
+            dismiss()
+            locationProvider.getLocation { location ->
+                locationSelectedListener(location)
+            }
+        } else {
+            permission.requestPermission(locationPermission)
```

Comment:
If it does need to request permission, will clicking the "use current location" button just appear to do nothing when clicked? Maybe we should add a failure handling of some sort.

Reply:
I think it'll ask the user for persmission and the popup will stay there. If they accept the permission, then they'll be able to use that button.